File: dashboard/src/types/inventory.ts
// Location types
export type LocationType = 'block' | 'floor' | 'room' | 'lab' | 'storage' | 'office';

export interface Location {
  id: string;
  tenantId: string;
  schoolId: string;
  parentId: string | null;
  name: string;
  locationType: LocationType;
  code: string;
  capacity: number;
  metadata?: Record<string, unknown>;
  active: boolean;
  createdAt: string;
  updatedAt: string;
  // Computed
  path?: string;
  deviceCount?: number;
}

export interface LocationTreeNode extends Location {
  children?: LocationTreeNode[];
}

export interface CreateLocationRequest {
  parentId?: string;
  name: string;
  locationType?: LocationType;
  code?: string;
  capacity?: number;
}

export interface UpdateLocationRequest {
  parentId?: string;
  name?: string;
  locationType?: LocationType;
  code?: string;
  capacity?: number;
  active?: boolean;
}

// Assignment types
export type AssignmentType = 'permanent' | 'temporary' | 'loan' | 'repair' | 'storage';

export interface DeviceAssignment {
  id: string;
  tenantId: string;
  deviceId: string;
  locationId: string | null;
  assignedToUser: string;
  assignmentType: AssignmentType;
  effectiveFrom: string;
  effectiveTo: string | null;
  notes: string;
  createdBy: string;
  createdAt: string;
  // Joined
  location?: Location;
}

export interface AssignDeviceRequest {
  locationId?: string;
  assignedToUser?: string;
  assignmentType?: AssignmentType;
  notes?: string;
}

// Group types
export type GroupType = 'manual' | 'location' | 'dynamic';

export interface GroupSelector {
  model?: string;
  lifecycle?: string[];
  locationType?: string;
  locationId?: string;
  schoolId?: string;
}

export interface DeviceGroup {
  id: string;
  tenantId: string;
  schoolId: string | null;
  name: string;
  description: string;
  groupType: GroupType;
  locationId: string | null;
  selector?: GroupSelector;
  policies?: Record<string, unknown>;
  active: boolean;
  createdAt: string;
  updatedAt: string;
  // Computed
  memberCount?: number;
}

export interface GroupMember {
  id: string;
  tenantId: string;
  groupId: string;
  deviceId: string;
  addedAt: string;
  addedBy: string;
}

export interface CreateGroupRequest {
  schoolId?: string;
  name: string;
  description?: string;
  groupType?: GroupType;
  locationId?: string;
  selector?: GroupSelector;
}

export interface AddGroupMembersRequest {
  deviceIds: string[];
}

// Device registration
export interface RegisterDeviceRequest {
  serial: string;
  assetTag?: string;
  model: string;
  make?: string;
  notes?: string;
  locationId?: string;
}

// Inventory device (enriched device view)
export interface InventoryDevice {
  id: string;
  tenantId: string;
  serial: string;
  assetTag: string;
  model: string;
  make: string;
  schoolId: string;
  lifecycle: string;
  enrolled: boolean;
  lastSeenAt?: string;
  createdAt: string;
  updatedAt: string;
  // Joined
  location?: Location;
  locationPath?: string;
  macAddresses?: string[];
  groups?: string[];
}

// Inventory summary
export interface InventorySummary {
  totalDevices: number;
  byStatus: Record<string, number>;
  byLocation: {
    assigned: number;
    unassigned: number;
  };
  byModel?: Record<string, number>;
}

// School inventory response
export interface SchoolInventoryResponse {
  school: {
    id: string;
    name: string;
  };
  summary: InventorySummary;
  devices: InventoryDevice[];
  locations: Location[];
}

// List responses
export interface LocationsListResponse {
  items: Location[];
}

export interface LocationTreeResponse {
  tree: LocationTreeNode[];
}

export interface AssignmentsListResponse {
  items: DeviceAssignment[];
}

export interface GroupsListResponse {
  items: DeviceGroup[];
}

export interface GroupMembersListResponse {
  items: GroupMember[];
}

// Constants for UI
export const LOCATION_TYPE_OPTIONS: Array<{ value: LocationType; label: string }> = [
  { value: 'block', label: 'Block' },
  { value: 'floor', label: 'Floor' },
  { value: 'room', label: 'Room' },
  { value: 'lab', label: 'Computer Lab' },
  { value: 'storage', label: 'Storage' },
  { value: 'office', label: 'Office' },
];

export const ASSIGNMENT_TYPE_OPTIONS: Array<{ value: AssignmentType; label: string }> = [
  { value: 'permanent', label: 'Permanent' },
  { value: 'temporary', label: 'Temporary' },
  { value: 'loan', label: 'On Loan' },
  { value: 'repair', label: 'In Repair' },
  { value: 'storage', label: 'In Storage' },
];

export const GROUP_TYPE_OPTIONS: Array<{ value: GroupType; label: string }> = [
  { value: 'manual', label: 'Manual' },
  { value: 'location', label: 'Location-based' },
  { value: 'dynamic', label: 'Dynamic' },
];

export const LOCATION_TYPE_ICONS: Record<LocationType, string> = {
  block: 'Building2',
  floor: 'Layers',
  room: 'DoorOpen',
  lab: 'Monitor',
  storage: 'Warehouse',
  office: 'Briefcase',
};

File: services/ims-api/internal/models/inventory.go
package models

import (
	"encoding/json"
	"time"
)

// LocationType represents the type of a location.
type LocationType string

const (
	LocationTypeBlock   LocationType = "block"
	LocationTypeFloor   LocationType = "floor"
	LocationTypeRoom    LocationType = "room"
	LocationTypeLab     LocationType = "lab"
	LocationTypeStorage LocationType = "storage"
	LocationTypeOffice  LocationType = "office"
)

// Location represents a physical location within a school (room, lab, block, etc.)
type Location struct {
	ID           string          `json:"id"`
	TenantID     string          `json:"tenantId"`
	SchoolID     string          `json:"schoolId"`
	ParentID     *string         `json:"parentId,omitempty"`
	Name         string          `json:"name"`
	LocationType LocationType    `json:"locationType"`
	Code         string          `json:"code"`
	Capacity     int             `json:"capacity"`
	Metadata     json.RawMessage `json:"metadata,omitempty"`
	Active       bool            `json:"active"`
	CreatedAt    time.Time       `json:"createdAt"`
	UpdatedAt    time.Time       `json:"updatedAt"`

	// Computed fields (not stored)
	Path        string `json:"path,omitempty"`        // Full hierarchical path: "Block A > Floor 1 > Lab 101"
	DeviceCount int    `json:"deviceCount,omitempty"` // Number of devices at this location
}

// AssignmentType represents the type of device assignment.
type AssignmentType string

const (
	AssignmentTypePermanent AssignmentType = "permanent"
	AssignmentTypeTemporary AssignmentType = "temporary"
	AssignmentTypeLoan      AssignmentType = "loan"
	AssignmentTypeRepair    AssignmentType = "repair"
	AssignmentTypeStorage   AssignmentType = "storage"
)

// DeviceAssignment represents a device's assignment to a location or user.
type DeviceAssignment struct {
	ID             string         `json:"id"`
	TenantID       string         `json:"tenantId"`
	DeviceID       string         `json:"deviceId"`
	LocationID     *string        `json:"locationId,omitempty"`
	AssignedToUser string         `json:"assignedToUser,omitempty"`
	AssignmentType AssignmentType `json:"assignmentType"`
	EffectiveFrom  time.Time      `json:"effectiveFrom"`
	EffectiveTo    *time.Time     `json:"effectiveTo,omitempty"` // nil = current
	Notes          string         `json:"notes,omitempty"`
	CreatedBy      string         `json:"createdBy,omitempty"`
	CreatedAt      time.Time      `json:"createdAt"`

	// Computed/joined fields
	Location *Location `json:"location,omitempty"`
}

// GroupType represents the type of device group.
type GroupType string

const (
	GroupTypeManual   GroupType = "manual"   // Manually managed membership
	GroupTypeLocation GroupType = "location" // Auto-membership by location
	GroupTypeDynamic  GroupType = "dynamic"  // Auto-membership by selector criteria
)

// GroupSelector defines criteria for dynamic group membership.
type GroupSelector struct {
	Model        string   `json:"model,omitempty"`        // Device model contains
	Lifecycle    []string `json:"lifecycle,omitempty"`    // Match any lifecycle status
	LocationType string   `json:"locationType,omitempty"` // Match location type
	LocationID   string   `json:"locationId,omitempty"`   // Match specific location
	SchoolID     string   `json:"schoolId,omitempty"`     // Match specific school
}

// DeviceGroup represents a group of devices for control/policy purposes.
type DeviceGroup struct {
	ID          string          `json:"id"`
	TenantID    string          `json:"tenantId"`
	SchoolID    *string         `json:"schoolId,omitempty"` // nil = tenant-wide
	Name        string          `json:"name"`
	Description string          `json:"description,omitempty"`
	GroupType   GroupType       `json:"groupType"`
	LocationID  *string         `json:"locationId,omitempty"` // For location-based groups
	Selector    *GroupSelector  `json:"selector,omitempty"`   // For dynamic groups
	Policies    json.RawMessage `json:"policies,omitempty"`   // Future: exam_mode, restrictions
	Active      bool            `json:"active"`
	CreatedAt   time.Time       `json:"createdAt"`
	UpdatedAt   time.Time       `json:"updatedAt"`

	// Computed fields
	MemberCount int `json:"memberCount,omitempty"`
}

// GroupMember represents a device's membership in a group.
type GroupMember struct {
	ID       string    `json:"id"`
	TenantID string    `json:"tenantId"`
	GroupID  string    `json:"groupId"`
	DeviceID string    `json:"deviceId"`
	AddedAt  time.Time `json:"addedAt"`
	AddedBy  string    `json:"addedBy,omitempty"`
}

// DeviceNetworkSnapshot is a cached copy of MAC address data from ssot-devices.
type DeviceNetworkSnapshot struct {
	TenantID      string     `json:"tenantId"`
	DeviceID      string     `json:"deviceId"`
	MACAddress    string     `json:"macAddress"`
	InterfaceType string     `json:"interfaceType"`
	IsPrimary     bool       `json:"isPrimary"`
	LastSeenAt    *time.Time `json:"lastSeenAt,omitempty"`
	SyncedAt      time.Time  `json:"syncedAt"`
}

// InventoryDevice represents a device with its inventory context.
type InventoryDevice struct {
	ID         string     `json:"id"`
	TenantID   string     `json:"tenantId"`
	Serial     string     `json:"serial"`
	AssetTag   string     `json:"assetTag"`
	Model      string     `json:"model"` // Denormalized from device_model
	Make       string     `json:"make"`
	SchoolID   string     `json:"schoolId"`
	Lifecycle  string     `json:"lifecycle"`
	Enrolled   bool       `json:"enrolled"`
	LastSeenAt *time.Time `json:"lastSeenAt,omitempty"`
	CreatedAt  time.Time  `json:"createdAt"`
	UpdatedAt  time.Time  `json:"updatedAt"`

	// Joined data
	Location     *Location `json:"location,omitempty"`
	LocationPath string    `json:"locationPath,omitempty"`
	MACAddresses []string  `json:"macAddresses,omitempty"`
	Groups       []string  `json:"groups,omitempty"` // Group IDs
}

// InventorySummary provides aggregate statistics for inventory views.
type InventorySummary struct {
	TotalDevices int            `json:"totalDevices"`
	ByStatus     map[string]int `json:"byStatus"`   // lifecycle status -> count
	ByLocation   map[string]int `json:"byLocation"` // "assigned" vs "unassigned"
	ByModel      map[string]int `json:"byModel,omitempty"`
}

// SchoolInventoryResponse is the response for GET /v1/schools/{schoolId}/inventory.
type SchoolInventoryResponse struct {
	School struct {
		ID   string `json:"id"`
		Name string `json:"name"`
	} `json:"school"`
	Summary   InventorySummary  `json:"summary"`
	Devices   []InventoryDevice `json:"devices"`
	Locations []Location        `json:"locations"`
}

// LocationTreeNode represents a location with its children for hierarchical views.
type LocationTreeNode struct {
	Location
	Children []LocationTreeNode `json:"children,omitempty"`
}
